### PR TITLE
CURA-4186 Add stack context for function evaluation

### PR DESCRIPTION
**Pull request by LipuFei**
CURA-4186

### CODE REVIEW

File: cura/Settings/ExtruderStack.py [hunk 1]
```diff
@@ -65,12 +65,14 @@ def getProperty(self, key: str, property_name: str, context: Optional[PropertyEv
...
-        if not super().getProperty(key, "settable_per_extruder"):
+        if not super().getProperty(key, "settable_per_extruder", context):
             result = self.getNextStack().getProperty(key, property_name, context)
             context.popContainer()
             return result
 
-        limit_to_extruder = super().getProperty(key, "limit_to_extruder")
+        limit_to_extruder = super().getProperty(key, "limit_to_extruder", context)
+        if limit_to_extruder is not None:
+            limit_to_extruder = str(limit_to_extruder)
```
Comment (Ghostkeeper) [inline cura/Settings/ExtruderStack.py]: I don't see why this is necessary?

File: cura/Settings/GlobalStack.py [hunk 2]
```diff
@@ -109,9 +109,11 @@ def getProperty(self, key: str, property_name: str, context: Optional[PropertyEv
         context.pushContainer(self)
 
         # Handle the "limit_to_extruder" property.
-        limit_to_extruder = super().getProperty(key, "limit_to_extruder")
+        limit_to_extruder = super().getProperty(key, "limit_to_extruder", context)
+        if limit_to_extruder is not None:
+            limit_to_extruder = str(limit_to_extruder)
```
Comment (Ghostkeeper) [inline cura/Settings/GlobalStack.py]: Why is this necessary?

File: cura/Settings/PerObjectContainerStack.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+from typing import Any, Optional
+
+from UM.Decorators import override
+from UM.Logger import Logger
+from UM.Settings.Interfaces import PropertyEvaluationContext
+from UM.Settings.ContainerStack import ContainerStack
+
+
+class PerObjectContainerStack(ContainerStack):
```
Comment (Ghostkeeper) [inline cura/Settings/PerObjectContainerStack.py]: This class should be located inside the per-object settings plug-in.

File: cura/Settings/PerObjectContainerStack.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    @override(ContainerStack)
+    def getProperty(self, key: str, property_name: str, context: Optional[PropertyEvaluationContext] = None) -> Any:
+        if context is None:
+            context = PropertyEvaluationContext()
+        context.pushContainer(self)
+
+        # this import has to be here otherwise there will be a circular import loop
+        from cura.CuraApplication import CuraApplication
+        global_stack = CuraApplication.getInstance().getGlobalContainerStack()
```
Comment (Ghostkeeper) [inline cura/Settings/PerObjectContainerStack.py]: Try `Application.getInstance().getGlobalContainerStack()`. You won't have a circular import then.

File: cura/Settings/PerObjectContainerStack.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        # Handle the "limit_to_extruder" property.
+        limit_to_extruder = super().getProperty(key, "limit_to_extruder", context)
+        if limit_to_extruder is not None:
+            limit_to_extruder = str(limit_to_extruder)
+
+        # if this stack has the limit_to_extruder "not overriden", use the original limit_to_extruder as the current
+        # limit_to_extruder, so the values retrieved will be from the perspective of the original limit_to_extruder
+        # stack.
+        if limit_to_extruder == "-1":
+            if 'original_limit_to_extruder' in context.context:
```
Comment (Ghostkeeper) [inline cura/Settings/PerObjectContainerStack.py]: Use double quotes for strings.

File: cura/Settings/PerObjectContainerStack.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                result = global_stack.extruders[str(limit_to_extruder)].getProperty(key, property_name, context)
+                if result is not None:
+                    context.popContainer()
+                    return result
+            else:
+                Logger.log("e", "Setting {setting} has limit_to_extruder but is not settable per extruder!", setting = key)
+
+        result = super().getProperty(key, property_name, context)
+        context.popContainer()
+        return result
```
Comment (Ghostkeeper) [inline cura/Settings/PerObjectContainerStack.py]: Shouldn't this limit to extruder handling just always be applied, rather than only in per-object stacks? I'd have put it in the main ContainerStack class. Or better yet, `InstanceContainer` even.

File: cura/Settings/PerObjectContainerStack.py [hunk 7]
```diff
@@ -17,6 +18,13 @@ def getProperty(self, key: str, property_name: str, context: Optional[PropertyEv
         from cura.CuraApplication import CuraApplication
         global_stack = CuraApplication.getInstance().getGlobalContainerStack()
 
+        # Return the user defined value if present, otherwise, evaluate the value according to the default routine.
+        if self.getContainer(0).hasProperty(key, property_name):
+            if self.getContainer(0)._instances[key].state == InstanceState.User:
+                result = super().getProperty(key, property_name, context)
+                context.popContainer()
+                return result
```
Comment (Ghostkeeper) [inline cura/Settings/PerObjectContainerStack.py]: User defined values are never formulas, so I don't see why this is necessary or why this is specific to per-object settings.

File: cura/Settings/PerObjectContainerStack.py [hunk 8]
```diff
@@ -51,3 +51,16 @@ def getProperty(self, key: str, property_name: str, context: Optional[PropertyEv
...
+        super().setNextStack(stack)
+
+        # trigger signal to re-evaluate all default settings
+        for key, instance in self.getContainer(0)._instances.items():
+            # only evaluate default settings
+            if instance.state != InstanceState.Default:
+                continue
+
+            self._collectPropertyChanges(key, "value")
+        self._emitCollectedPropertyChanges()
```
Comment (Ghostkeeper) [inline cura/Settings/PerObjectContainerStack.py]: Settings on the stack should not get updated. They are fixed from the moment that they are added, and should not get re-evaluated after that.

File: cura/Settings/PerObjectContainerStack.py (hunk 5 above)
Comment (LipuFei) [reply cura/Settings/PerObjectContainerStack.py]: Fixed in 37414e7bc7c85ee85ce47289a3268f1214bda2b3

File: cura/Settings/PerObjectContainerStack.py (hunk 4 above)
Comment (LipuFei) [reply cura/Settings/PerObjectContainerStack.py]: Fixed in b84500f742db7ffb9352065027024b0a3f913ec6
